### PR TITLE
Add test leg to the PR build to run libraries tests on Android emulators

### PR DESCRIPTION
**Pull request by premun**
This is a fork of https://github.com/dotnet/runtime/pull/37476

Adds the Android test leg that will execute selected BCL tests on an Android emulator using the `Ubuntu.1804.Amd64.Open` queue.

To prepare the Helix job, we are using the Arcade Helix SDK, more specifically the [XHarness Helix SDK](https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/Readme.md) which uses the [XHarness tool](https://github.com/dotnet/xharness) to control the test run.

### CODE REVIEW

File: eng/pipelines/libraries/build-job.yml [hunk 1]
```diff
@@ -104,6 +106,13 @@ jobs:
                 sourceFolder: $(Build.SourcesDirectory)/artifacts/bin/testhost
                 targetFolder: $(Build.ArtifactStagingDirectory)/artifacts/bin/testhost
 
+            - ${{ if eq(parameters.runtimeFlavor, 'mono') }}:
+              - task: CopyFiles@2
+                displayName: Prepare lib-runtime-packs folder to publish
+                inputs:
+                  sourceFolder: $(Build.SourcesDirectory)/artifacts/bin/microsoft.netcore.app.runtime.ios-x64
```
Comment (premun) [inline eng/pipelines/libraries/build-job.yml:113]: This will need to be fixed

File: src/libraries/sendtohelix.proj [hunk 2]
```diff
@@ -97,7 +102,28 @@
...
+      <XHarnessAppFolderToTest Include="$([System.IO.Directory]::GetDirectories('$(TestArchiveTestsRoot)', '*.app', System.IO.SearchOption.AllDirectories))">
+        <Targets Condition="'$(TargetArchitecture)' == 'arm'">ios-device</Targets>
+        <Targets Condition="'$(TargetArchitecture)' == 'arm64'">ios-device</Targets>
+        <Targets Condition="'$(TargetArchitecture)' == 'x64'">ios-simulator-64</Targets>
+        <Targets Condition="'$(TargetArchitecture)' == 'x86'">ios-simulator-32</Targets>
+      </XHarnessAppFolderToTest>
+    </ItemGroup>
+
+    <ItemGroup Condition="'$(TargetOS)' == 'Android'">
+      <_apks Include="$(TestArchiveTestsRoot)**/*.apk" />
```
Comment (safern) [inline src/libraries/sendtohelix.proj:115]: Why introduce a separate itemgroup? We can just do this directly in `XHarnessPackageToTest`

File: eng/install-native-dependencies.sh [hunk 3]
```diff
@@ -15,6 +15,16 @@ elif [ "$1" = "OSX" ] || [ "$1" = "tvOS" ] || [ "$1" = "iOS" ]; then
...
+       exit 1;
+    fi
+    if [ -d "${ANDROID_OPENSSL_AAR}" ]; then
+        exit 0;
+    fi
+    OPENSSL_VER=1.1.1g-alpha-1
+    curl https://maven.google.com/com/android/ndk/thirdparty/openssl/${OPENSSL_VER}/openssl-${OPENSSL_VER}.aar -L --output /tmp/openssl.zip
+    unzip /tmp/openssl.zip -d "${ANDROID_OPENSSL_AAR}" && rm -rf /tmp/openssl.zip
 else
     echo "Must pass \"Linux\", \"tvOS\", \"iOS\" or \"OSX\" as first argument."
```
Comment (safern) [inline eng/install-native-dependencies.sh:29]: Update this as it now supports Android.

File: eng/install-native-dependencies.sh [hunk 4]
```diff
@@ -15,6 +15,16 @@ elif [ "$1" = "OSX" ] || [ "$1" = "tvOS" ] || [ "$1" = "iOS" ]; then
     if [ "$?" != "0" ]; then
         exit 1;
     fi
+elif [ "$1" = "Android" ]; then
+    if [ -z "${ANDROID_OPENSSL_AAR}" ]; then
+       exit 1;
```
Comment (safern) [inline eng/install-native-dependencies.sh:20]: Print out an error saying that this env var needs to be set?

File: src/libraries/sendtohelix.proj (hunk 2 above)
Comment (premun) [reply src/libraries/sendtohelix.proj:115]: I was actually testing this locally and was unable to access the `%(Filename)` metadata when I had the `Include` directly in `XHarnessPackageToTest`. I debugged the binlogs yesterday for an hour and don't understand why, but it was the case. Even `%(Identity)` wasn't set..

File: eng/pipelines/common/xplat-setup.yml [hunk 5]
```diff
@@ -99,12 +99,12 @@ jobs:
     ${{ if eq(parameters.jobParameters.pool, '') }}:
       pool:
         # Public Linux Build Pool
-        ${{ if and(in(parameters.osGroup, 'Linux', 'FreeBSD'), eq(variables['System.TeamProject'], 'public')) }}:
+        ${{ if and(in(parameters.osGroup, 'Android', 'Linux', 'FreeBSD'), eq(variables['System.TeamProject'], 'public')) }}:
```
Comment (safern) [inline eng/pipelines/common/xplat-setup.yml:102]: Maybe leave this as it was to keep using the hosted pools for Android builds? I noticed this will really slow down the machine provisioning.

File: eng/pipelines/common/xplat-setup.yml [hunk 6]
```diff
@@ -99,12 +99,12 @@ jobs:
...
           name:  NetCorePublic-Pool
           queue: BuildPool.Ubuntu.1604.Amd64.Open
 
         # Official Build Linux Pool
-        ${{ if and(in(parameters.osGroup, 'Linux', 'FreeBSD'), ne(variables['System.TeamProject'], 'public')) }}:
+        ${{ if and(in(parameters.osGroup, 'Android', 'Linux', 'FreeBSD'), ne(variables['System.TeamProject'], 'public')) }}:
```
Comment (safern) [inline eng/pipelines/common/xplat-setup.yml:107]: ditto.

File: eng/Versions.props [hunk 7]
```diff
@@ -107,6 +107,7 @@
     <!-- Testing -->
     <MicrosoftNETTestSdkVersion>16.7.0-preview-20200608-10</MicrosoftNETTestSdkVersion>
     <MicrosoftDotNetXHarnessTestRunnersXunitVersion>1.0.0-prerelease.20305.1</MicrosoftDotNetXHarnessTestRunnersXunitVersion>
+    <MicrosoftDotNetXHarnessCLIVersion>1.0.0-prerelease.20308.1</MicrosoftDotNetXHarnessCLIVersion>
```
Comment (safern) [inline eng/Versions.props:110]: We should add this to `Versions.Details.xml` in order to get auto-updates for it.

File: eng/pipelines/common/global-build-job.yml [hunk 8]
```diff
@@ -54,7 +60,7 @@ jobs:
     - ${{ if eq(parameters.isOfficialBuild, true) }}:
       - template: /eng/pipelines/common/restore-internal-tools.yml
 
-    - ${{ if in(parameters.osGroup, 'OSX', 'iOS', 'tvOS') }}:
+    - ${{ if or(in(parameters.osGroup, 'OSX', 'iOS', 'tvOS'), and(eq(parameters.osGroup, 'Android'), eq(parameters.isOfficialBuild, false))) }}:
```
Comment (safern) [inline eng/pipelines/common/global-build-job.yml:63]: @directhex @premun do we need only to run tests or also for the final shipping product?

Comment (premun) [reply eng/pipelines/common/global-build-job.yml:63]: I have no idea about this, just copied it from @steveisok

File: eng/pipelines/common/global-build-job.yml [hunk 9]
```diff
@@ -54,6 +54,11 @@ jobs:
         ${{ if ne(parameters.isOfficialBuild, true) }}:
           value: ''
 
+      - ${{ if and(eq(parameters.osGroup, 'Android'), eq(parameters.isOfficialBuild, false)) }}:
+        - name: ANDROID_OPENSSL_AAR
+          value: /tmp/openssl-android
+
+      - ${{ parameters.variables }}
```
Comment (safern) [inline eng/pipelines/common/global-build-job.yml:61]: This is wrong, we're doing the same with the foreach below, basically propagating the parameters.variables that are an array.

File: eng/pipelines/common/xplat-setup.yml [hunk 10]
```diff
@@ -103,12 +103,12 @@ jobs:
     ${{ if eq(parameters.jobParameters.pool, '') }}:
       pool:
         # Public Linux Build Pool
-        ${{ if and(in(parameters.osGroup, 'Linux', 'FreeBSD'), eq(variables['System.TeamProject'], 'public')) }}:
+        ${{ if and(in(parameters.osGroup, 'Linux', 'FreeBSD', 'Android'), eq(variables['System.TeamProject'], 'public')) }}:
```
Comment (safern) [inline eng/pipelines/common/xplat-setup.yml:106]: Let's revert this change. Android builds inside a docker container so let's use the hosted pools which is effectively by leaving this as it was.

File: eng/pipelines/common/xplat-setup.yml [hunk 11]
```diff
@@ -103,12 +103,12 @@ jobs:
...
           name:  NetCorePublic-Pool
           queue: BuildPool.Ubuntu.1604.Amd64.Open
 
         # Official Build Linux Pool
-        ${{ if and(in(parameters.osGroup, 'Linux', 'FreeBSD'), ne(variables['System.TeamProject'], 'public')) }}:
+        ${{ if and(in(parameters.osGroup, 'Linux', 'FreeBSD', 'Android'), ne(variables['System.TeamProject'], 'public')) }}:
```
Comment (safern) [inline eng/pipelines/common/xplat-setup.yml:111]: Ditto.

File: eng/pipelines/runtime.yml [hunk 12]
```diff
@@ -288,11 +288,9 @@ jobs:
     buildConfig: Release
     runtimeFlavor: mono
     platforms:
-    - Android_x64
     - Android_arm
     - tvOS_arm64
     - iOS_arm
-    - iOS_x64
```
Comment (safern) [inline eng/pipelines/runtime.yml:295]: This is probably not intentional right?

File: eng/pipelines/runtime.yml [hunk 13]
```diff
@@ -326,7 +325,7 @@ jobs:
       testGroup: innerloop
       nameSuffix: AllSubsets_Mono
       buildArgs: -s mono+libs+installer+libs.tests -c $(_BuildConfig) /p:ArchiveTests=true
-      timeoutInMinutes: 120
+      timeoutInMinutes: 240
```
Comment (safern) [inline eng/pipelines/runtime.yml:328]: Is this because building the android tests takes a long time, or to actually running them?

File: eng/testing/AppleRunnerTemplate.sh [hunk 14]
```diff
@@ -38,15 +35,25 @@ while true; do
...
+    sudo kill "$pid"
+fi
+open -a "$simulator_app"
+export XHARNESS_OUT="$EXECUTION_DIR/xharness-output"
 
-XHARNESS_OUT="$EXECUTION_DIR/xharness-output"
-
-dotnet xharness ios test --app="$APP_BUNDLE" \
-    --targets=$TARGET \
+dotnet xharness ios test \
```
Comment (safern) [inline eng/testing/AppleRunnerTemplate.sh:49]: Let's match this what we did for `Browser` where the `RunCommands` are defined in `tests.mobile.targets` that way we can preserve xunit arguments like `-notrait Category=IgnoreForCI` or others like `Outerloop`.

File: eng/testing/tests.mobile.targets [hunk 15]
```diff
@@ -72,6 +83,7 @@
     <Error Condition="!Exists('$(MicrosoftNetCoreAppRuntimePackRidDir)')" Text="MicrosoftNetCoreAppRuntimePackRidDir=$(MicrosoftNetCoreAppRuntimePackRidDir) doesn't exist" />
     <Error Condition="('$(TargetArchitecture)' == 'arm64' or '$(TargetArchitecture)' == 'arm') and '$(DevTeamProvisioning)' == ''"
            Text="'DevTeamProvisioning' needs to be set for device builds. Set it to 'UBF8T346G9' if you're part of the Microsoft team account." />
+    <Error Condition="'$(TestArchiveTestsDir)' == ''" Text="TestArchiveTestsDir property to archive the test folder must be set." />
```
Comment (safern) [inline eng/testing/tests.mobile.targets:86]: This property should only be set if `ArchiveTests == true`

File: src/libraries/sendtohelixhelp.proj [hunk 16]
```diff
@@ -91,7 +91,7 @@
     <HelixCommand>$(HelixCommand) /p:LocalPackagesPath="%HELIX_CORRELATION_PAYLOAD%\packages\"</HelixCommand>
   </PropertyGroup>
 
-  <ItemGroup Condition="'$(HelixCommand)' == '' and '$(MonoEnvOptions)' != ''">
+  <ItemGroup Condition="'$(HelixCommand)' == '' and '$(MonoEnvOptions)' != '' and '$(TargetOS)' != 'Android'">
```
Comment (safern) [inline src/libraries/sendtohelixhelp.proj:94]: I don't see what value adding this condition brings us since `MonoEnvOptions` should be empty for Android runs, right?

**Comment by safern** [comment]
You can ignore my comments now. Just leaving them as I reviewed already and so that we don't forget to cleanup before merging.

File: eng/pipelines/runtime.yml (hunk 13 above)
Comment (directhex) [reply eng/pipelines/runtime.yml:328]: At the time, there was only one booted Helix VM running Android tests, which take a combined ~3 hours on one host

File: eng/pipelines/runtime.yml (hunk 12 above)
Comment (directhex) [reply eng/pipelines/runtime.yml:295]: Pretty much every change in runtime.yml is a temporary hack

File: eng/pipelines/common/xplat-setup.yml (hunk 10 above)
Comment (safern) [reply eng/pipelines/common/xplat-setup.yml:106]: Yeah, you're right forgot about the size constraint for Android. I think we should definitely improve that though

File: eng/testing/tests.mobile.targets [hunk 17]
```diff
@@ -9,14 +9,20 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <RunScriptCommands Include="TEST_NAME=$(TestProjectName)" />
```
Comment (akoeplinger) [inline eng/testing/tests.mobile.targets:12]: doesn't seem to be used in the .sh, can be removed

File: eng/testing/tests.mobile.targets [hunk 18]
```diff
@@ -9,14 +9,20 @@
...
     <RunScriptCommands Include="TARGET_ARCH=$(TargetArchitecture)" />
+    <RunScriptCommands Include="APK=$EXECUTION_DIR/bin/$(TestProjectName).apk" Condition="'$(TargetOS)' == 'Android'"/>
```
Comment (akoeplinger) [inline eng/testing/tests.mobile.targets:14]: same here, this can be moved down inline to `RunScriptCommand`

File: src/libraries/sendtohelixhelp.proj [hunk 19]
```diff
@@ -163,7 +163,19 @@
     <Message Importance="High" Text="Using WorkItemArchiveWildCard: $(WorkItemArchiveWildCard)" />
     <Message Importance="High" Text="Using Timeout: $(_timeoutSpan)" />
 
-    <ItemGroup>
+    <ItemGroup Condition="'$(TargetOS)' == 'Android'">
+      <_apks Include="$(TestArchiveTestsRoot)**/*.apk" />
```
Comment (akoeplinger) [inline src/libraries/sendtohelixhelp.proj:167]: why do we need this extra variable?

File: eng/pipelines/runtime.yml [hunk 20]
```diff
@@ -321,7 +321,7 @@ jobs:
...
```
Comment (safern) [inline eng/pipelines/runtime.yml:324]: Do we still need such a long timeout?

File: src/libraries/sendtohelixhelp.proj (hunk 19 above)
Comment (safern) [reply src/libraries/sendtohelixhelp.proj:167]: This is in order to be able to use `%(Filename)` when setting the metadata... it is an MSBuild bug which I can't find at the moment.

File: eng/pipelines/runtime.yml (hunk 20 above)
Comment (directhex) [reply eng/pipelines/runtime.yml:324]: We did a couple of days ago, but it looks like I'm running this PR often enough that Helix is allowing more than one Android test machine to provision so I can lower it again

Comment (safern) [reply eng/pipelines/runtime.yml]: Thanks, we can always bump it up if needed when a lot of PRs start to send helix jobs if we can't get more capacity.

File: src/libraries/sendtohelixhelp.proj (hunk 19 above)
Comment (premun) [reply src/libraries/sendtohelixhelp.proj:167]: I will put a comment there about it

**Comment by steveisok** [comment]
Note, sometimes tests may hit https://github.com/dotnet/runtime/issues/38213.  As noted in the issue, my suspicion is that sometimes (frequency unknown to me) emulators may run out of memory.

**Comment by directhex** [comment]
Android... passed without complaint?

`  Job 8023d290-c15e-46dc-a314-dbea9bb05de6 on Ubuntu.1804.Amd64.Android.Open is completed with 204 finished work items.`

**Comment by premun** [comment]
I merged to get a new attempt and this time it unfortunately [timed out](
https://dev.azure.com/dnceng/public/_build/results?buildId=810975&view=logs&j=b6e5d97f-d4f8-506b-9e47-e3e32385754a&t=e50f4001-4d55-54a5-1d7b-f04ae1a4c3e4&l=51)

**Comment by premun** [comment]
I merged to get a new attempt and this time it [timed out](
https://dev.azure.com/dnceng/public/_build/results?buildId=810975&view=logs&j=b6e5d97f-d4f8-506b-9e47-e3e32385754a&t=e50f4001-4d55-54a5-1d7b-f04ae1a4c3e4&l=51).

However..
- The job itself succeeded (all tests passed)
- The overall job took **1 hour 20 minutes** from Queued to Finished
- The job was stopped after **1 hour 3 minutes** because the AzDO stage was limited to 2 hours
- The overall runtime is **2 hours 31 minutes**
- The whole job was processed on **2 agents only**, if it was 3 agents, it would finish just under 1 hour
- **Longest** suite took **6 minutes** to complete
- **Average** runtime was **47 seconds** per suite

**Comment by directhex** [comment]
Walltime-related failures should sort themselves, after a few goes to convince Helix to spin up more VMs

**Comment by directhex** [comment]
What remaining complaints do we have on this PR? It seems pleasantly green

**Comment by premun** [comment]
We need to bump the number of cores in the Android queue to support the load. Currently, it was set to 2 which hardly manages to finish 1 build (204 work items) in time

**Comment by MattGal** [comment]
> We need to bump the number of cores in the Android queue to support the load. Currently, it was set to 2 which hardly manages to finish 1 build (204 work items) in time.
> 
> @MattGal I am not sure about this - who has the authority to decide whether we can afford this and we should fund this?

Until we do the other issue that gives the emulators more ram, workitems WILL randomly fail. It's literally just writing a config file while deploying the artifact.

As for more VMs, @AlitzelMendez, @sunandabalu,  and @ulisesh are who I talk to about this but if you keep using any given queue, it will slowly be given more capacity.

**Comment by sunandabalu** [comment]
> > We need to bump the number of cores in the Android queue to support the load. Currently, it was set to 2 which hardly manages to finish 1 build (204 work items) in time.
> > @MattGal I am not sure about this - who has the authority to decide whether we can afford this and we should fund this?
> 
> Until we do the other issue that gives the emulators more ram, workitems WILL randomly fail. It's literally just writing a config file while deploying the artifact.
> 
> As for more VMs, @AlitzelMendez, @sunandabalu, and @ulisesh are who I talk to about this but if you keep using any given queue, it will slowly be given more capacity.

We are rolling out the feature to calculate max capacity for a queue based on the current queue depth everytime scaling happens instead of depending on a static value on 9/23.

**Comment by directhex** [comment]
emulator RAM PR is merged.

**Comment by premun** [comment]
Unfortunately, we didn't release yesterday, so the auto-scaler is not out yet. Also the images with the RAM changes are not out neither. I think next attempt is next Wednesday unfortunately

**Comment by MattGal** [comment]
> Unfortunately, we didn't release yesterday, so the auto-scaler is not out yet. Also the images with the RAM changes are not out neither. I think next attempt is next Wednesday unfortunately

With the autoscaler bugs resolved, rollout is looking possible tomorrow (Thursday)

**Comment by directhex** [comment]
Well now the autoscaler got rebooted, things are looking healthier

File: eng/pipelines/common/global-build-job.yml [hunk 21]
```diff
@@ -65,7 +69,7 @@ jobs:
     - ${{ if eq(parameters.isOfficialBuild, true) }}:
       - template: /eng/pipelines/common/restore-internal-tools.yml
 
-    - ${{ if in(parameters.osGroup, 'OSX', 'iOS', 'tvOS') }}:
+    - ${{ if or(in(parameters.osGroup, 'OSX', 'iOS', 'tvOS'), eq(parameters.osGroup, 'Android')) }}:
```
Comment (safern) [inline eng/pipelines/common/global-build-job.yml:72]: instead just add `Android` to the `in` list, to make the condition simpler.

**Review by safern — APPROVED** [review]
Other than one comment looks good to me.

File: eng/pipelines/common/global-build-job.yml (hunk 21 above)
Comment (premun) [reply eng/pipelines/common/global-build-job.yml:72]: I also added 'Android` to line 85 with the `Disk Usage after Build` build step
